lib/AST/SwiftNameTranslation.cpp

```diff
@@ -256,19 +259,29 @@ struct ObjCTypeWalker : TypeWalker {
 } // namespace
 
 bool swift::cxx_translation::isObjCxxOnly(const ValueDecl *VD) {
-  ObjCTypeWalker walker;
+  ObjCTypeWalker walker{VD->getASTContext()};
   VD->getInterfaceType().walk(walker);
   return walker.hadObjCType;
 }
 
-bool swift::cxx_translation::isObjCxxOnly(const clang::Decl *D) {
+bool swift::cxx_translation::isObjCxxOnly(const clang::Decl *D,
+                                          const ASTContext &ctx) {
   // By default, we import all modules in Obj-C++ mode, so there is no robust
-  // way to tell if something is coming from an Obj-C module. The best
-  // approximation is to check if something is a framework module as most of
-  // those are written in Obj-C. Ideally, we want to add `requires ObjC` to all
-  // ObjC modules and respect that here to make this completely precise.
-  return D->getASTContext().getLangOpts().ObjC &&
-         D->getOwningModule()->isPartOfFramework();
+  // way to tell if something is coming from an Obj-C module. Use the
+  // requirements and the language options to check if we should actually
+  // consider the module to have ObjC constructs.
+  const auto &langOpts = D->getASTContext().getLangOpts();
+  auto clangModule = D->getOwningModule()->getTopLevelModule();
+  bool requiresObjC = false;
+  for (auto req : clangModule->Requirements)
+    if (req.RequiredState && req.FeatureName == "ObjC")
```

Comment:
I think it's spelled `objc` in lowercase: https://clang.llvm.org/docs/Modules.html#requires-declaration